### PR TITLE
Feature/save image issue

### PR DESCRIPTION
**Pull request by Mee-gu**
fix save image issue

### CODE REVIEW

File: cocos/2d/CCRenderTexture.cpp [hunk 1]
```diff
@@ -300,10 +301,16 @@ void RenderTexture::beginWithClear(float r, float g, float b, float a, float dep
...
+    Director::getInstance()->getRenderer()->clear(_clearFlags, _clearColor, _clearDepth, _clearStencil, _globalZOrder);
 }
 
 void RenderTexture::clear(float r, float g, float b, float a)
+{
+    this->beginWithClear(r, g, b, a);
+    this->end();
+}
+
+void RenderTexture::clearColor(float r, float g, float b, float a)
```
Comment (minggo) [inline cocos/2d/CCRenderTexture.cpp]: It is not needed.

File: cocos/renderer/backend/Texture.h [hunk 2]
```diff
@@ -22,6 +24,7 @@ class Texture : public Ref
 {
 public:
     virtual void updateSamplerDescriptor(const SamplerDescriptor &sampler) = 0;
+    virtual void getBytes(int x, int y, int width, int height, TextureFormat format, bool flipImage, std::function<void(const unsigned char* imageRGBA)> callback) = 0;
```
Comment (minggo) [inline cocos/renderer/backend/Texture.h]: imageRGBA is not suitable

Comment (minggo) [reply cocos/renderer/backend/Texture.h]: format is not needed, just return data is enough

File: cocos/renderer/backend/opengl/CommandBufferGL.cpp [hunk 3]
```diff
@@ -118,6 +118,18 @@ void CommandBufferGL::beginFrame()
...
+    auto commandType = command->getType();
+    switch (commandType) {
+        case RenderCommand::Type::SYCHRONIZED_CALLBACK_COMMAND:
+            static_cast<SynchronizedCallbackCommand*>(command)->execute();
+            break;
+        default:
+            break;
+    }
+}
+
```
Comment (minggo) [inline cocos/renderer/backend/opengl/CommandBufferGL.cpp]: should delete

File: cocos/renderer/backend/metal/TextureMTL.mm [hunk 4]
```diff
@@ -182,4 +183,55 @@ bool convertData(uint8_t* src, unsigned int length, TextureFormat format, uint8_
...
+    [commandEncoder synchronizeResource:copiedTexture];
+#endif
+    [commandEncoder endEncoding];
+    
+    [commandBuffer addCompletedHandler:^(id<MTLCommandBuffer> commandBuffer) {
+        MTLRegion region = MTLRegionMake2D(0, 0, width, height);
+        unsigned char* imageRGBA = new unsigned char[width * height *4];
+        [copiedTexture getBytes:imageRGBA bytesPerRow:width*4 fromRegion:region mipmapLevel:0];
+        
+        if(flipImage)
```
Comment (minggo) [inline cocos/renderer/backend/metal/TextureMTL.mm]: flipImage -> !flipImage

File: cocos/renderer/backend/opengl/CommandBufferGL.h [hunk 5]
```diff
@@ -40,8 +40,8 @@ class CommandBufferGL final : public CommandBuffer
     
     virtual void setScissorRect(bool isEnabled, float x, float y, float width, float height) override;
     virtual void setDepthStencilState(DepthStencilState* depthStencilState) override;
+    virtual void setCallBackCommand(RenderCommand* comand) override;
```
Comment (minggo) [inline cocos/renderer/backend/opengl/CommandBufferGL.h]: should delete

File: cocos/2d/CCRenderTexture.cpp [hunk 6]
```diff
@@ -389,93 +391,64 @@ bool RenderTexture::saveToFile(const std::string& fileName, Image::Format format
...
-    if (image)
-    {
-        image->saveToFile(filename, !isRGBA);
-    }
-    if(_saveFileCallback)
-    {
-        _saveFileCallback(this, filename);
-    }
-    CC_SAFE_DELETE(image);
+    auto callbackFunc = [&](Image* image, const std::string& filePath, bool isRGBA){
```
Comment (minggo) [inline cocos/2d/CCRenderTexture.cpp]: filename, isRGBA don't have to be variables

File: tests/cpp-tests/Classes/NewRendererTest/NewRendererTest.cpp [hunk 7]
```diff
@@ -29,18 +29,18 @@ USING_NS_CC;
...
     ADD_TEST_CASE(CaptureScreenTest);
     ADD_TEST_CASE(CaptureNodeTest);
-    ADD_TEST_CASE(BugAutoCulling);
-    ADD_TEST_CASE(RendererBatchQuadTri);
-    // ADD_TEST_CASE(RendererUniformBatch); // TODO shouldn't call OpenGL API directly
-    // ADD_TEST_CASE(RendererUniformBatch2);
+//    ADD_TEST_CASE(BugAutoCulling);
+//    ADD_TEST_CASE(RendererBatchQuadTri);
+//    // ADD_TEST_CASE(RendererUniformBatch); // TODO shouldn't call OpenGL API directly
+//    // ADD_TEST_CASE(RendererUniformBatch2);
```
Comment (minggo) [inline tests/cpp-tests/Classes/NewRendererTest/NewRendererTest.cpp]: please revert debug codes

**Comment by Mee-gu** [comment]
update review

**Pull request merged**